filebeat/input/file/state.go

```diff
@@ -81,25 +83,36 @@ func (s *States) findPrevious(newState State) (int, State) {
 }
 
 // Cleanup cleans up the state array. All states which are older then `older` are removed
-func (s *States) Cleanup(older time.Duration) {
+func (s *States) Cleanup() {
+
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
 	for i, state := range s.states {
 
-		// File wasn't seen for longer then older -> remove state
-		if time.Since(state.LastSeen) > older {
-			logp.Debug("prospector", "State removed for %s because of older: %s", state.Source)
-			s.states = append(s.states[:i], s.states[i+1:]...)
+		// -1 means ttl is disabled
+		if state.Ttl == -1 {
+			continue
 		}
-	}
 
+		// States is older then ttl
+		if int64(time.Since(state.Timestamp).Nanoseconds()/int64(time.Microsecond)) > state.Ttl {
```

Comment:
I just converted and now realised that it seems like this breaks the json conversion. Any trick here?

```
	TTL         time.Duration     `json:"ttl"`
```